### DISCUSSION TITLE
fitz.open() - Python keeps file open

### DISCUSSION THREAD

**Discussion by gosforth** [pymupdf/PyMuPDF#1547] [Looking for help] [answered]
Hi,
I open pdf file:
` doc = fitz.open(pfile)`
At the end I close it
` doc.close()`
And I check if is closed:
` isclosed = doc.is_closed`
But another process says this file is kept by Python.
In previous version that worked fine.

How can I open this file only for reading?

EDIT
I've just came back to version 1.18.19 and that work well. Bug?

**Comment by JorjMcKie** [comment]
that is a problem in the current version
fixed in next v1.19.5

**Reply by gosforth** [reply]
Oh My, I spent 5 hours fighting with that :-)
What do you recommend - keeping 1.18.19 or move to 1.19.3?

Anyway, it will be better to open file:

```
with fitz.open(ifile) as doc:
 ...
```
and Python close it itself
Or without **with** and close file with **doc.close()**?

**Reply by JorjMcKie** [reply]
You could have asked earlier - others have no such inhibitions 😉
If you have Linux or Mac OSX, you will find current pre-version wheels [here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse). If Windows drop me a note.

**Reply by gosforth** [reply]
Thanks a lot for your help.
Concerning my philosophical question above? Use 'with' or clean up after myself? :-)
Is there option to open file only for reading?

**Reply by JorjMcKie** [reply]
one of the reasons for the issue in v1.19.4 was the effort to unify all the different Document destruction places.
Then I forgot that the C reference counter specifically for documents maybe > 1. So I had to build in (again) a loop that drops the document while any references exist.
So in short: starting v1.19.5 that makes no difference any more. You also can use `doc = None` or `del doc` with the same effect.

**Reply by AlexeyPopkov** [reply]
> You could have asked earlier - others have no such inhibitions 😉 If you have Linux or Mac OSX, you will find current pre-version wheels [here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse). If Windows drop me a note.

I have this problem on Windows.

**Reply by JorjMcKie** [reply]
@AlexeyPopkov - please tell me Python version and bitness

**Reply by AlexeyPopkov** [reply]
> @AlexeyPopkov - please tell me Python version and bitness

Python 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)]

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.19.5-cp39-cp39-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/7918124/PyMuPDF-1.19.5-cp39-cp39-win_amd64.zip)

Unzip and execute `py -3.9 -m pip install --force-reinstall PyMuPDF-1.19.5-cp39-cp39-win_amd64.whl`.

**Reply by AlexeyPopkov** [reply]
> [PyMuPDF-1.19.5-cp39-cp39-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/7918124/PyMuPDF-1.19.5-cp39-cp39-win_amd64.zip)
> 
> Unzip and execute `py -3.9 -m pip install --force-reinstall PyMuPDF-1.19.5-cp39-cp39-win_amd64.whl`.

Thank you! The problem disappeared.

**Comment by SnejPro** [comment]
I have this issue with 1.22.1

**Reply by Sythys** [reply]
I have this issue with 1.23.17

**Reply by julian-smith-artifex-com** [reply]
This issue is fixed in 1.23.18.

**Comment by yonglee7015** [comment]
I have another question about fitz.open() with html. 
I use flask to develop a web app.
<input style="height:50px;width:600px;font-size:20px;" type="file" id="file" size="30" name="file" accept="application/pdf" multiple 

files = request.files.getlist("file")
fitz.open can't open file in files. , fitz can't read the path html get. But pdfplumber can open this path.
Why fitz limit this ?

**Reply by JorjMcKie** [reply, marked as answer]
Your code is minimal and I am not sure where the files really exist you want to open.
But maybe this is an explanation:
To open a document with PyMuPDF, then this must either be a file existing on local drives, or it must be a memory area in formta of the respective document.
PyMuPDF does not access the internet as part of `fitz.open()`. You must first download any internet-resident file e.g., to a memory area (not necessarily to disk).
That maybe explains it.

**Reply by yonglee7015** [reply]
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file

with this element,  fitz can't open selected files directly.

**Reply by JorjMcKie** [reply]
> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file
> 
> with this element, fitz can't open selected files directly.

That is because it is embedded in a HTML prompt.
